### PR TITLE
Fix FastText model reading of unsupported modes from Facebook's FastText

### LINKED ISSUE #3179: Gensim's FastText model reads in unsupported modes from Facebook's FastText
In gensim/models/fasttext.py:

```python
    model = FastText(
        vector_size=m.dim,
        vector_size=m.dim,
        window=m.ws,
        window=m.ws,
        epochs=m.epoch,
        epochs=m.epoch,
        negative=m.neg,
        negative=m.neg,
        # FIXME: these next 2 lines read in unsupported FB FT modes (loss=3 softmax or loss=4 onevsall,
        # or model=3 supervised) possibly creating inconsistent gensim model likely to fail later. Displaying
        # clear error/warning with explanatory message would be far better - even if there might be some reason
        # to continue with the load - such as providing read-only access to word-vectors trained those ways. (See:
        # https://github.com/facebookresearch/fastText/blob/2cc7f54ac034ae320a9af784b8145c50cc68965c/src/args.h#L19
        # for FB FT mode definitions.)
        hs=int(m.loss == 1),
        hs=int(m.loss == 1),
        sg=int(m.model == 2),
        sg=int(m.model == 2),
        bucket=m.bucket,
        bucket=m.bucket,
        min_count=m.min_count,
        min_count=m.min_count,
        sample=m.t,
        sample=m.t,
        min_n=m.minn,
        min_n=m.minn,
        max_n=m.maxn,
        max_n=m.maxn,
    )
```

### PR DESCRIPTION
**Pull request by Tewatia5355**
Fixes #3179: Added error handling.

### CODE REVIEW

**Comment by SagarDollin** [comment]
Hey @Tewatia5355 @karan121bhukar , 
Similar to your PR , even [my PR](https://github.com/RaRe-Technologies/gensim/pull/3220)  passes all the checks except for the build wheels. I have cloned the project on a windows system. Any idea how to fix this?

**Comment by Tewatia5355** [comment]
> Hey @Tewatia5355 @karan121bhukar ,
> Similar to your PR , even [my PR](https://github.com/RaRe-Technologies/gensim/pull/3220) passes all the checks except for the build wheels. I have cloned the project on a windows system. Any idea how to fix this?

Hey there, I'm still figuring out why it didn't worked!! Will tell you if I get some leads :)

**Comment by SagarDollin** [comment]
> Hey there, I'm still figuring out why it didn't worked!! Will tell you if I get some leads :)

I think this is because the file [build-wheels](https://github.com/RaRe-Technologies/gensim/commit/b3e820bcc708b95cbedab9627d03a6c34af4ea8c#diff-14b51e9e6321caf5d23b2253f2415daf475fdaa417c0aa8fe896c3a460f2d023) . A change was made 13 days ago where they added username. The error we are getting is also related to username:
image

tagging @mpenkov : Should we create an issue for this. I observed that in the previous pull requests this issue wasn't seen. Infact the newer PR seems to be having more checks than previous one's.

**Comment by mpenkov** [comment]
Thank you for letting me know. I'll disable wheel building for PRs.

**Comment by SagarDollin** [comment]
> Thank you for letting me know. I'll disable wheel building for PRs.

Thank you, does this mean we have to create another pull request to pass the checks?

**Comment by mpenkov** [comment]
No, there's no need for that.

File: gensim/models/fasttext.py [hunk 1]
```diff
@@ -807,6 +807,13 @@ def _load_fasttext_format(model_file, encoding='utf-8', full_model=True):
     with utils.open(model_file, 'rb') as fin:
         m = gensim.models._fasttext_bin.load(fin, encoding=encoding, full_model=full_model)
 
+    # Here we are checking for unsupported FB FT Modes
+    if m.loss != 1 and m.loss != 2:
```
Comment (piskvorky) [inline gensim/models/fasttext.py:811]:
```suggestion
    if m.loss not in (1, 2):
```

File: gensim/models/fasttext.py [hunk 2]
```diff
@@ -807,6 +807,13 @@ def _load_fasttext_format(model_file, encoding='utf-8', full_model=True):
...
+        raise ValueError("Loss paramter value can be either 1 (for Hierarchical Softmax) or 2 (for Negative Sampling)")
+    elif m.model != 1 and m.model != 2:
```
Comment (piskvorky) [inline gensim/models/fasttext.py:813]:
```suggestion
    elif m.model not in (1, 2):
```

File: gensim/models/fasttext.py [hunk 3]
```diff
@@ -807,6 +807,13 @@ def _load_fasttext_format(model_file, encoding='utf-8', full_model=True):
     with utils.open(model_file, 'rb') as fin:
         m = gensim.models._fasttext_bin.load(fin, encoding=encoding, full_model=full_model)
 
+    # Here we are checking for unsupported FB FT Modes
+    if m.loss != 1 and m.loss != 2:
+        raise ValueError("Loss paramter value can be either 1 (for Hierarchical Softmax) or 2 (for Negative Sampling)")
```
Comment (piskvorky) [inline gensim/models/fasttext.py:812]:
```suggestion
        raise ValueError("The fasttext `loss` parameter must be either 1 (for Hierarchical Softmax) or 2 (for Negative Sampling)")
```

File: gensim/models/fasttext.py [hunk 4]
```diff
@@ -807,6 +807,13 @@ def _load_fasttext_format(model_file, encoding='utf-8', full_model=True):
...
+    elif m.model != 1 and m.model != 2:
+        raise ValueError(
+            "Model parameter value can be either 1 (for Continous Bag of Words model) or 2 (for Skip-gram model)")
```
Comment (piskvorky) [inline gensim/models/fasttext.py:815]:
```suggestion
            "The fasttext `model` parameter must be either 1 (for Continuous Bag of Words model) or 2 (for Skip-gram model)")
```

File: gensim/similarities/fastss.c [hunk 5]
```diff
@@ -0,0 +1,8416 @@
+/* Generated by Cython 0.29.23 */
```
Comment (piskvorky) [inline gensim/similarities/fastss.c]: Why is this here? Please remove from the PR.

File: gensim/models/fasttext.py (hunk 1 above)
Comment (piskvorky) [reply gensim/models/fasttext.py:811]: According to https://github.com/RaRe-Technologies/gensim/pull/3223, negative sampling is `loss=0`, not `loss=2`. Can you please double check? What's the correct value?

**Review by piskvorky — CHANGES_REQUESTED** [review]
.

File: gensim/models/fasttext.py [hunk 6]
```diff
@@ -807,6 +807,13 @@ def _load_fasttext_format(model_file, encoding='utf-8', full_model=True):
     with utils.open(model_file, 'rb') as fin:
         m = gensim.models._fasttext_bin.load(fin, encoding=encoding, full_model=full_model)
 
+    # Here we are checking for unsupported FB FT Modes
```
Comment (piskvorky) [inline gensim/models/fasttext.py:810]:
```suggestion
    # Check for unsupported Facebook FastText modes.
```

**Review by piskvorky — CHANGES_REQUESTED** [review]

**Review by SagarDollin — APPROVED** [review]

**Comment by piskvorky** [comment]
@SagarDollin can you address the above please? We're planning a new release soon, it'd be great to include your fix.

**Comment by SagarDollin** [comment]
> @SagarDollin can you address the above please? We're planning a new release soon, it'd be great to include your fix.

Hi. I have approved the changes you requested. I'll also give some time for the comments you mentioned. However Im currently in between some other work and I'll get back to this within 2-3 days. 

By addressing this did you mean only approving the requested changes or to address all the comments?

Thanks,
Sagar B Dollin

**Review by SagarDollin — APPROVED** [review]

**Review by SagarDollin — APPROVED** [review]

**Review by SagarDollin — APPROVED** [review]

**Comment by piskvorky** [comment]
> By addressing this did you mean only approving the requested changes or to address all the comments?

"All the comments" – although I only see one: `loss=0` vs `loss=2`. Is there anything else unresolved?

The rest are cosmetic language changes, to be simply accepted.

**Comment by SagarDollin** [comment]
> > By addressing this did you mean only approving the requested changes or to address all the comments?
> 
> "All the comments" – although I only see one: `loss=0` vs `loss=2`. Is there anything else unresolved?
> 
> The rest are cosmetic language changes, to be simply accepted.

By all the other comments I meant these comments on PR #3220 :
https://github.com/RaRe-Technologies/gensim/pull/3220#pullrequestreview-887942010

**Comment by piskvorky** [comment]
Hm, is that necessary? #3220 is a separate PR.

**Comment by SagarDollin** [comment]
> Hm, is that necessary? #3220 is a separate PR. Is it blocking this PR in some way?

No not at all. I'll take a look at this one first. 

Thanks..

**Comment by piskvorky** [comment]
@SagarDollin can you finish this please? We'd like to release ASAP. Thanks!

File: gensim/models/fasttext.py (hunk 1 above)
Comment (SagarDollin) [reply gensim/models/fasttext.py:811]: Hi,
I checked it. In the parameters docstring in the models/_fasttext_bin, that I have copied here:
```
"""Holds data loaded from the Facebook binary.
Parameters

.
.
neg : int
    If non-zero, indicates that the model uses negative sampling.
loss : int
    If equal to 1, indicates that the model uses hierarchical sampling.
model : int
    If equal to 2, indicates that the model uses skip-grams.
"""
````
This suggests that `m.neg==0` means no negative sampling, and `m.neg == n` (where n is a positive integer) means negative sampling is being used. 
So I think if we are checking for negative sampling, the condition should be `if m.neg == 0` or `if m.neg != 0` based on what we want to do. From what I have read from the  #3223 I gather that negative sampling and hierarchical softmax are supported in loss. so we shouldn't really need to check for negative sampling i.e, `if m.neg != 0` (not really required)

I have made a list of what's supported and what's not based on #3223 . Please let me know if they are right, ill make changes as per that:

Supported:
a. Loss

    1. Negative Sampling
    2. Hierarchical Softmax
 b. Model 

    1. CBOW
    2. Skip-gram

Unsupported:
a. Loss:
 1. OneVsAll_Loss

b. Model:
  1. Supervised
Let me know if there are any inconsistencies in my understanding. I too want to help you complete this asap.

Comment (piskvorky) [reply gensim/models/fasttext.py:811]: I don't really understand that list. If the check is `== 0` or `!= 0`, why would we check for `== 1` and `== 2`?

I don't know the fasttext parameters, so can you please check the fasttext docs (not gensim) for what the parameters **actually** are? And then make changes to match that. Thanks.

Comment (SagarDollin) [reply gensim/models/fasttext.py:811]: > Sorry, I don't really understand that list. If the check is `== 0` or `!= 0`, why would we check for `== 1` and `== 2`?

Hi yes you are right. We shouldn't be checking `==1` or `==2` in this case also if we are checking for negative sampling we must not, use `m.loss` at all we must check for `m.neg`. (Just to clarify the above ==1 and ==2 was not done by me, this PR was created by @Tewatia5355, Im just trying to figure out what are the losses and models not supported in fasttext. )
> 
> I don't know the fasttext parameters, so don't ask me :) Can you please check the fasttext docs or code (not gensim) for what the parameters **actually** are? And then make changes here to match that. Thanks.

Hi, yes sure I can investigate that and ill create a PR accordingly. Also if you can provide the link for fasttext docs that would be great. (just to be sure that I'm referring the right docs). Thanks!

Comment (piskvorky) [reply gensim/models/fasttext.py:811]: OK, please do. This PR seems incorrect so I'm removing the "Next release" tag – we need to fix it first, make the code correct.

**Comment by piskvorky** [comment]
The code in this PR seems incorrect (see https://github.com/RaRe-Technologies/gensim/pull/3222#discussion_r851611240) and its author went MIA, so closing here. But fixing the issue is still desirable.
